src/Features/Core/Portable/Diagnostics/CodeAnalysisDiagnosticAnalyzerService.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host.Mef;
+using System.Threading.Tasks;
+using System.Composition;
+using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    [ExportWorkspaceServiceFactory(typeof(ICodeAnalysisDiagnosticAnalyzerService), ServiceLayer.Default), Shared]
+    internal sealed class CodeAnalysisDiagnosticAnalyzerServiceFactory : IWorkspaceServiceFactory
+    {
+        private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+        private readonly IDiagnosticsRefresher _diagnosticsRefresher;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CodeAnalysisDiagnosticAnalyzerServiceFactory(
+            IDiagnosticAnalyzerService diagnosticAnalyzerService,
+            IDiagnosticsRefresher diagnosticsRefresher)
+        {
+            _diagnosticAnalyzerService = diagnosticAnalyzerService;
+            _diagnosticsRefresher = diagnosticsRefresher;
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+            => new CodeAnalysisDiagnosticAnalyzerService(_diagnosticAnalyzerService, _diagnosticsRefresher, workspaceServices.Workspace);
+
+        private sealed class CodeAnalysisDiagnosticAnalyzerService : ICodeAnalysisDiagnosticAnalyzerService
+        {
+            private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+            private readonly IDiagnosticsRefresher _diagnosticsRefresher;
+            private readonly Workspace _workspace;
+            private readonly ConcurrentSet<ProjectId> _analyzedProjectIds = new();
+
+            public CodeAnalysisDiagnosticAnalyzerService(
+                IDiagnosticAnalyzerService diagnosticAnalyzerService,
+                IDiagnosticsRefresher diagnosticsRefresher,
+                Workspace workspace)
+            {
+                _diagnosticAnalyzerService = diagnosticAnalyzerService;
+                _diagnosticsRefresher = diagnosticsRefresher;
+                _workspace = workspace;
+
+                _workspace.WorkspaceChanged += OnWorkspaceChanged;
```

Comment:
don't love this (As we want to get away from workspace's eventing themselves.  but this is ok for now.